### PR TITLE
Simulated writing to characteristics

### PR DESCRIPTION
**Pull request by bartoszwilk**

### CODE REVIEW

File: lib/blemulator/internal/bridge/platform_to_dart_bridge.dart [hunk 1]
```diff
@@ -73,22 +79,29 @@ class PlatformToDartBridge {
...
+          (e) => <String, dynamic>{
+            SimulationArgumentName.uuid: e.uuid,
+            SimulationArgumentName.id: e.id,
+            SimulationArgumentName.characteristics: e
+                .characteristics()
+                .map(
+                  (e) => <String, dynamic>{
+                    SimulationArgumentName.id: e.id,
+                    SimulationArgumentName.uuid: e.uuid,
+                    //TODO expand model to send all fields
```
Comment (pawelByszewski) [inline lib/blemulator/internal/bridge/platform_to_dart_bridge.dart]: Why it is not expanded in this PR?

File: example/lib/test_scenarios/peripheral_test_operations.dart [hunk 2]
```diff
@@ -113,14 +113,46 @@ class PeripheralTestOperations {
...
-        Uint8List.fromList([1]),
+        SensorTagTemperatureUuids.temperatureDataCharacteristic,
+        Uint8List.fromList([0, 200, 40, 0]),
         false);
 
-    log("Turned on temperature update");
+    log("Written to characteristic");
+  }
+
+  Future<void> writeCharacteristicForService() async {
```
Comment (pawelByszewski) [inline example/lib/test_scenarios/peripheral_test_operations.dart]: IMHO it should be introduced in the develop and then merged to `blesimulator/develop`, if not then when will those changes appear in the develop?

File: ios/Classes/Constants/ArgumentKey.m [hunk 3]
```diff
@@ -23,7 +23,7 @@ @implementation ArgumentKey
 NSString * const ARGUMENT_KEY_SERVICE_ID = @"serviceId";
 NSString * const ARGUMENT_KEY_CHARACTERISTIC_UUID = @"characteristicUuid";
 NSString * const ARGUMENT_KEY_CHARACTERISTIC_IDENTIFIER = @"characteristicIdentifier";
-NSString * const ARGUMENT_KEY_BYTES = @"bytes";
+NSString * const ARGUMENT_KEY_VALUE = @"value";
```
Comment (pawelByszewski) [inline ios/Classes/Constants/ArgumentKey.m]: are you able to test if it works?

File: ios/Classes/Constants/ArgumentKey.h [hunk 4]
```diff
@@ -21,7 +21,7 @@ extern NSString * const ARGUMENT_KEY_SERVICE_UUID;
 extern NSString * const ARGUMENT_KEY_SERVICE_ID;
 extern NSString * const ARGUMENT_KEY_CHARACTERISTIC_UUID;
 extern NSString * const ARGUMENT_KEY_CHARACTERISTIC_IDENTIFIER;
-extern NSString * const ARGUMENT_KEY_BYTES;
+extern NSString * const ARGUMENT_KEY_VALUE;
```
Comment (pawelByszewski) [inline ios/Classes/Constants/ArgumentKey.h]: are you able to test if it works?

File: ios/Classes/Constants/ArgumentKey.m (hunk 3 above)
Comment (TomoLV) [reply ios/Classes/Constants/ArgumentKey.m]: I'm getting `Error Domain=CBATTErrorDomain Code=3 "Writing is not permitted."` when running test with real device SensorTag but I suppose it's fine, because test scenario is written for simulated devices, right?

File: example/lib/test_scenarios/peripheral_test_operations.dart (hunk 2 above)
Comment (bartoszwilk) [reply example/lib/test_scenarios/peripheral_test_operations.dart]: Done. https://github.com/Polidea/FlutterBleLib/pull/225

File: lib/blemulator/internal/bridge/platform_to_dart_bridge.dart (hunk 1 above)
Comment (bartoszwilk) [reply lib/blemulator/internal/bridge/platform_to_dart_bridge.dart]: fixed

File: ios/Classes/Constants/ArgumentKey.m (hunk 3 above)
Comment (TomoLV) [reply ios/Classes/Constants/ArgumentKey.m]: Tested again - error does not occur anymore

File: android/src/main/java/com/polidea/blemulator/bridging/DartMethodCaller.java [hunk 5]
```diff
@@ -373,4 +354,97 @@ public void notImplemented() {
...
+                arguments,
+                new MethodChannel.Result() {
+                    @Override
+                    public void success(@Nullable Object characteristicJsonObject) {
+                        onSuccessCallback.onSuccess(characteristicJsonDecoder.decode((Map<String, Object>) characteristicJsonObject));
+                    }
+
+                    @Override
+                    public void error(String errorCode, @Nullable String message, @Nullable Object bleErrorJsonObject) {
+                        onErrorCallback.onError(new BleError(BleErrorCode.UnknownError, message, 0)); //TODO Add proper error parsing here
```
Comment (pawelByszewski) [inline android/src/main/java/com/polidea/blemulator/bridging/DartMethodCaller.java]: please add an issue for the TODO(or link existing one) or implement a proper solution

File: example/lib/test_scenarios/peripheral_test_operations.dart [hunk 6]
```diff
@@ -67,14 +67,33 @@ class PeripheralTestOperations {
...
+        SensorTagTemperatureUuids.temperatureDataCharacteristic);
+    log("Read temperature value ${_convertToTemperature(readValue.value)}C");
+  }
+
+  Future<void> writeCharacteristicForPeripheral() async {
+    log("Write to characteristic via peripheral");
+    await peripheral.writeCharacteristic(
+        SensorTagTemperatureUuids.temperatureService,
+        SensorTagTemperatureUuids.temperatureDataCharacteristic,
+        Uint8List.fromList([0, 200, 40, 0]),
```
Comment (pawelByszewski) [inline example/lib/test_scenarios/peripheral_test_operations.dart]: 1. why did you changes `Uint8List.fromList([1]),` to `Uint8List.fromList([0, 200, 40, 0]),`?
2. Why it is not introduced into the example on develop?

File: ios/Classes/FlutterBleLibPlugin.m [hunk 7]
```diff
@@ -315,7 +315,7 @@ - (void)writeCharacteristicForService:(FlutterMethodCall *)call result:(FlutterR
 
 - (void)writeCharacteristic:(FlutterMethodCall *)call result:(FlutterResult)result {
     [_manager writeCharacteristic:[call.arguments[ARGUMENT_KEY_CHARACTERISTIC_IDENTIFIER] doubleValue]
-                      valueBase64:[self base64encodedStringFromBytes:call.arguments[ARGUMENT_KEY_BYTES]]
+                      valueBase64:[self base64encodedStringFromBytes:call.arguments[ARGUMENT_KEY_VALUE]]
```
Comment (mikolak) [inline ios/Classes/FlutterBleLibPlugin.m]: Why simulated writing to characteristics touches non-simulated iOS bridging library?

File: lib/blemulator/internal/bridge/platform_to_dart_bridge.dart [hunk 8]
```diff
@@ -73,22 +79,25 @@ class PlatformToDartBridge {
...
+            call.arguments[ArgumentName.id] as String);
+    dynamic mapped = services
+        .map(
+          (service) => <String, dynamic>{
+            Metadata.serviceUuid: service.uuid,
+            Metadata.serviceId: service.id,
+            SimulationArgumentName.characteristics: service
+                .characteristics()
+                .map(
+                  (characteristic) => _convertToMap(call.arguments[ArgumentName.id], characteristic, null),
```
Comment (mikolak) [inline lib/blemulator/internal/bridge/platform_to_dart_bridge.dart]: Run autoformatter

File: lib/blemulator/internal/bridge/platform_to_dart_bridge.dart [hunk 9]
```diff
@@ -123,26 +136,76 @@ class PlatformToDartBridge {
...
-            characteristic));
+              arguments[SimulationArgumentName.deviceIdentifier],
+              characteristic,
+              arguments[SimulationArgumentName.value],
+            ));
   }
 
   Map<String, dynamic> _convertToMap(
-          String peripheralId, CharacteristicResponse response) =>
+    String peripheralId,
```
Comment (mikolak) [inline lib/blemulator/internal/bridge/platform_to_dart_bridge.dart]: Why is peripheralId always sent to platform, even in discovery?

File: lib/internal/bridge/characteristics_mixin.dart [hunk 10]
```diff
@@ -79,7 +79,7 @@ mixin CharacteristicsMixin on FlutterBLE {
         MethodName.writeCharacteristicForIdentifier,
         <String, dynamic>{
           ArgumentName.characteristicIdentifier: characteristicIdentifier,
-          ArgumentName.bytes: bytes,
+          ArgumentName.value: bytes,
```
Comment (mikolak) [inline lib/internal/bridge/characteristics_mixin.dart]: Why are there changes in FlutterBleLib?

File: lib/blemulator/internal/simulation_manager.dart [hunk 11]
```diff
@@ -45,77 +45,4 @@ class SimulationManager extends SimulationManagerBaseWithErrorChecks
     _scanSubscriptions.forEach((subscription) => subscription.cancel());
     _scanSubscriptions.clear();
   }
-
-  Future<CharacteristicResponse> _readCharacteristicForIdentifier(
```
Comment (mikolak) [inline lib/blemulator/internal/simulation_manager.dart]: Where all this went?

File: example/lib/test_scenarios/peripheral_test_operations.dart (hunk 6 above)
Comment (bartoszwilk) [reply example/lib/test_scenarios/peripheral_test_operations.dart]: Actually it was changed having only simulator example in mind. It would throw an error as data characteristic is not writable in real sensor tag. 
For now I will just leave writing "1" to characteristic config, but fully working example for characteristics I/O operations will be introduced in characteristics monitoring PR.

File: lib/blemulator/simulation/simulated_characteristic.dart [hunk 12]
```diff
@@ -34,7 +34,10 @@ class SimulatedCharacteristic {
   Future<Uint8List> read() async =>
       isReadable ? _value : Future.error("This characteristic is not readable");
 
-  void write(Uint8List value) {
+  Future<void> write(Uint8List value) async {
+    if (!isWritableWithResponse || !isWritableWithoutResponse) {
+      return Future.error("This characteristic is not writeable");
```
Comment (mikolak) [inline lib/blemulator/simulation/simulated_characteristic.dart]: Perhaps the checks should be done by simulation manager in error_mixin or whatever it was called? Either way: create a proper BleError here, don't throw a string.

File: lib/internal/constants.dart [hunk 13]
```diff
@@ -87,7 +87,7 @@ abstract class ArgumentName {
   static const String characteristicUuid = "characteristicUuid";
   static const String characteristicIdentifier = "characteristicIdentifier";
 
-  static const String bytes = "bytes";
+  static const String value = "value";
```
Comment (mikolak) [inline lib/internal/constants.dart]: Why this change and why in this PR?

File: ios/Classes/FlutterBleLibPlugin.m (hunk 7 above)
Comment (bartoszwilk) [reply ios/Classes/FlutterBleLibPlugin.m]: I decided to change this name while implementing this feature instead of providing inconsistent naming conventions (in one place characteristic value param would be called bytes and in another place it would be called value). Moreover I didn't want to continue using "bytes" name because, as we already discussed, its wrong.

File: lib/blemulator/internal/simulation_manager.dart (hunk 11 above)
Comment (bartoszwilk) [reply lib/blemulator/internal/simulation_manager.dart]: It was a redundancy

File: lib/blemulator/internal/bridge/platform_to_dart_bridge.dart (hunk 9 above)
Comment (bartoszwilk) [reply lib/blemulator/internal/bridge/platform_to_dart_bridge.dart]: For the sake of consistency. We've already discussed it, so Im resolving this conversation.

File: lib/internal/constants.dart (hunk 13 above)
Comment (bartoszwilk) [reply lib/internal/constants.dart]: I created a separate PR for FlutterBleLib https://github.com/Polidea/FlutterBleLib/pull/232

File: lib/internal/bridge/characteristics_mixin.dart (hunk 10 above)
Comment (bartoszwilk) [reply lib/internal/bridge/characteristics_mixin.dart]: I created a separate PR for FlutterBleLib https://github.com/Polidea/FlutterBleLib/pull/232

File: ios/Classes/FlutterBleLibPlugin.m (hunk 7 above)
Comment (bartoszwilk) [reply ios/Classes/FlutterBleLibPlugin.m]: I created a separate PR for FlutterBleLib https://github.com/Polidea/FlutterBleLib/pull/232

File: lib/blemulator/simulation/simulated_characteristic.dart (hunk 12 above)
Comment (bartoszwilk) [reply lib/blemulator/simulation/simulated_characteristic.dart]: 1. We decided to throw these errors from characteristic_mixin because now user loses this error handling when he forget to call super.write() in characteristic's overridden "write" method.
2. I will also add proper BleError for that case instead of returning raw string
3. We don't simulate writing with or without response. I created a separate issue for that (https://github.com/Polidea/FlutterBleLib/issues/234)

File: lib/blemulator/internal/bridge/platform_to_dart_bridge.dart [hunk 14]
```diff
@@ -123,26 +137,76 @@ class PlatformToDartBridge {
...
+        )
+        .then((characteristic) => _convertToMap(
+              arguments[SimulationArgumentName.deviceIdentifier],
+              characteristic,
+              arguments[SimulationArgumentName.value],
+            ));
+  }
+
+  Future<dynamic> _writeCharacteristicForService(MethodCall call) async {
+    Map<dynamic, dynamic> arguments = call.arguments;
```
[the remark this answers is not in the document]
Comment (bartoszwilk) [reply lib/blemulator/internal/bridge/platform_to_dart_bridge.dart]: Because then I get an error: _InternalLinkedHashMap<dynamic, dynamic>' is not a subtype of type 'Map<String, dynamic>'

File: lib/blemulator/internal/bridge/platform_to_dart_bridge.dart [hunk 15]
```diff
@@ -123,26 +137,76 @@ class PlatformToDartBridge {
...
+        )
+        .then((characteristicResponse) => _convertToMap(
+              arguments[SimulationArgumentName.deviceIdentifier],
+              characteristicResponse,
+              arguments[SimulationArgumentName.value],
+            ));
+  }
+
+  Future<dynamic> _writeCharacteristicForDevice(MethodCall call) async {
+    Map<dynamic, dynamic> arguments = call.arguments;
```
[the remark this answers is not in the document]
Comment (bartoszwilk) [reply lib/blemulator/internal/bridge/platform_to_dart_bridge.dart]: Because then I get an error: _InternalLinkedHashMap<dynamic, dynamic>' is not a subtype of type 'Map<String, dynamic>'

File: lib/blemulator/internal/bridge/platform_to_dart_bridge.dart [hunk 16]
```diff
@@ -123,26 +137,76 @@ class PlatformToDartBridge {
...
         )
+        .then((characteristicResponse) => _convertToMap(
+              arguments[SimulationArgumentName.deviceIdentifier],
+              characteristicResponse.characteristic,
+              characteristicResponse.value,
+            ));
+  }
+
+  Future<dynamic> _writeCharacteristicForIdentifier(MethodCall call) async {
+    Map<dynamic, dynamic> arguments = call.arguments;
```
[the remark this answers is not in the document]
Comment (bartoszwilk) [reply lib/blemulator/internal/bridge/platform_to_dart_bridge.dart]: Because then I get an error: _InternalLinkedHashMap<dynamic, dynamic>' is not a subtype of type 'Map<String, dynamic>'

File: lib/blemulator/internal/simulation_manager_mixins/characteristics_mixin.dart [hunk 17]
```diff
@@ -22,9 +22,7 @@ mixin CharacteristicsMixin on SimulationManagerBase {
     SimulatedCharacteristic targetCharacteristic =
         getCharacteristic(characteristicIdentifier);
 
-    if (targetCharacteristic == null)
-      return Future.error("Characteristic not found");
-
+    _throwErrorIfNull(targetCharacteristic);
```
[the remark this answers is not in the document]
Comment (bartoszwilk) [reply lib/blemulator/internal/simulation_manager_mixins/characteristics_mixin.dart]: Every exception that is thrown within "async" function is wrapped in Future and returned as a result of this function.
Screenshot 2019-10-14 at 17 22 34

File: example/lib/test_scenarios/peripheral_test_operations.dart [hunk 18]
```diff
@@ -76,14 +76,14 @@ class PeripheralTestOperations {
   }
 
   Future<void> writeCharacteristicForPeripheral() async {
-    log("Write to characteristic via peripheral");
+    log("Turning on temperature update");
```
Comment (mikolak) [inline example/lib/test_scenarios/peripheral_test_operations.dart]: `... via peripheral` 😄

File: lib/blemulator/internal/simulation_manager_mixins/characteristics_mixin.dart [hunk 19]
```diff
@@ -38,7 +38,7 @@ mixin CharacteristicsMixin on SimulationManagerBase {
     SimulatedCharacteristic targetCharacteristic = targetPeripheral
         .getCharacteristicForService(serviceUuid, characteristicUUID);
 
-    _throwErrorIfNull(targetCharacteristic);
+    _throwErrorIfNull(targetCharacteristic, characteristicUUID);
```
Comment (mikolak) [inline lib/blemulator/internal/simulation_manager_mixins/characteristics_mixin.dart]: What if targetPeripheral is null?

File: lib/blemulator/internal/simulation_manager_mixins/characteristics_mixin.dart [hunk 20]
```diff
@@ -59,7 +59,7 @@ mixin CharacteristicsMixin on SimulationManagerBase {
       if (characteristic != null) break peripheralsLoop;
     }
 
-    _throwErrorIfNull(targetCharacteristic);
+    _throwErrorIfNull(targetCharacteristic, characteristicUUID);
```
Comment (mikolak) [inline lib/blemulator/internal/simulation_manager_mixins/characteristics_mixin.dart]: This even works? targetCharacteristic is never set.

File: lib/blemulator/internal/simulation_manager_mixins/characteristics_mixin.dart [hunk 21]
```diff
@@ -89,7 +89,7 @@ mixin CharacteristicsMixin on SimulationManagerBase {
     SimulatedCharacteristic targetCharacteristic = targetPeripheral
         .getCharacteristicForService(serviceUuid, characteristicUUID);
 
-    _throwErrorIfNull(targetCharacteristic);
+    _throwErrorIfNull(targetCharacteristic, characteristicUUID);
```
Comment (mikolak) [inline lib/blemulator/internal/simulation_manager_mixins/characteristics_mixin.dart]: What if targetPeripheral is null?

File: lib/blemulator/internal/simulation_manager_mixins/characteristics_mixin.dart (hunk 19 above)
Comment (bartoszwilk) [reply lib/blemulator/internal/simulation_manager_mixins/characteristics_mixin.dart]: StateError from _peripherals.values.firstWhere() will be thrown and passed as a result of this "_readCharacteristicForDevice" method. There is already a task about checking if the device that we want to perform BLE operation on is known and if we are connected to it.

File: lib/blemulator/internal/simulation_manager_mixins/characteristics_mixin.dart (hunk 20 above)
Comment (bartoszwilk) [reply lib/blemulator/internal/simulation_manager_mixins/characteristics_mixin.dart]: It was a bug which was fixed in monitoring characteristics branch. I fixed it here too.
